Catch problematic warning during pnbd optimization, retry with bad observations removed

nickresnick opened the pull request:
While investigating convergence issues for the lifetimes pnbd, I found that when the following warning hits:
```
RuntimeWarning: invalid value encountered in subtract
  numpy.max(numpy.abs(fsim[0] - fsim[1:])) <= fatol):
```
it implies the likelihood is infinite (or in some cases nan) which causes the optimization to get stuck at the current parameter values.

This update catches the numpy warning as an actual RuntimeWarning, finds the individual entries that caused the likelihood to be infinite, and retries optimization with those entries removed.

I tested this out on the database export dates that caused brooks and golftown to produce the above warning, and with the update both sites converged successfully

Commit:
Catch problematic warning during pnbd optimization, retry

While investigating convergence issues for the lifetimes pnbd, I found that when the following warning hits:

RuntimeWarning: invalid value encountered in subtract
  numpy.max(numpy.abs(fsim[0] - fsim[1:])) <= fatol):

it implies the likelihood is infinite (or in some cases nan) which causes the optimization to get stuck at the current parameter values.

This update catches the numpy warning as an actual RuntimeWarning, finds the individual entries that caused the likelihood to be infinite, and retries optimization with those entries removed.

Kristjansson commented on lifetimes/fitters/pareto_nbd_fitter.py:
```diff
@@ -525,19 +528,64 @@ def _fit(
+                nan_indices = np.where(np.isnan(likelihoods))
+                bad_indices = np.concatenate((inf_indices[0], nan_indices[0]))
+                print(f"Number of bad indices: {bad_indices.size}")
+                print(f"Problematic entries: {f[bad_indices]}, {r[bad_indices]}, {t[bad_indices]}")
```
These will be entirely numeric, right? Specifically, is there any chance we're logging PII here?

nickresnick replied:
correct, no PII

Kristjansson commented on lifetimes/fitters/pareto_nbd_fitter.py:
```diff
@@ -525,19 +528,64 @@ def _fit(
+
+                minimizing_function_args = (f, r, t, weights, reg_param)
+
+                output = minimize(
```
Should we make this a recursive call to `_fit` so we have the option to recover more than once?

nickresnick replied:
This isn't a traditional retry mechanism, in that once we've removed the violating rows we should run the optimization once, but I don't think there's any benefit in attempting to retry again

Kristjansson replied:
So if we remove the rows, retry, and run into nan/infs again we'll fail. I'm not sure if this is the right mental model, but I was picturing a dataset with 2 outliers where one is much more of an outlier than the other. It seems reasonable that the large outlier would generate a nan, get removed, but then after a few iterations the smaller outlier would also over/under flow.

Do you think that's unlikely to happen? Or that if we get in a situation like that we'd want to fail and investigate? Or that it's just not worth the level of effort to handle multiple retries at the moment? For context, I'm fine if any of these is the case, just trying to get a good understanding of the problem.

nickresnick replied:
Ah got it. It's a good thought. I do agree it seems that theoretically the params could move in a direction that causes some other row to throw nans/infs, though I haven't seen it in practice. I can make this more robust tho

Kristjansson replied:
Ok cool, glad we're on the same page.

Using a recursive call seemed like a cheap way to make this more robust, but if there's more complexity than that it might not be worth it until we have evidence we need it.

Commit:
Raise error if more than 10k rows removed

Normally I'd expect no more than 5-10 rows to be removed, but given that we're usually fitting on 500k rows, 10k seems like a reasonable cutoff that would imply something is systematically wrong

Commit:
Add recursive try_optimize fn with retry limit

Kristjansson reviewed (approved):
looks good to me!
